### ISSUE TITLE
Tiles on the same layer as an entity in pixel mode need special treatment.

### ISSUE THREAD

**Issue by seisatsu** [seisatsu/Driftwood#162]
If a bush is on the same layer as the player, and they stand in front of it, they render on top of it, as expected. However if they stand behind it, they still render on top of it! While this makes perfect sense programmatically, it makes no sense aesthetically. Some special treatment is needed.

Suggested solution: While an entity's xy position is overlapping any part of the lower half of a tile on the same layer, the entity draws on top. Otherwise, it draws underneath. Once an entity's y position is further north than the halfway point of the tile, it "passes through".

The only other clean way I can think of to fix this issue is to not allow entities to overlap nowalk tiles at all in pixel mode, and I think we already decided against that.

**Comment by pmer** [comment]
This reminds me of [Paper Mario style graphics](https://youtu.be/XZnTmo7Y5ng?t=432).

Wouldn't a game would just want to keep tiles and entities on separate layers?

Although there is still the problem of entities overlapping each other which we might want to solve by sorting entities by Y-position before rendering.

**Comment by pmer** [comment]
One idea that might solve both entity-tile and entity-entity overlaps would be the following.

Instead of drawing all entities on the current layer in one go, intersperse them with the drawing of tile rows.

I'm picturing something like this (ignore the differences between this & existing Driftwood code):

```python
# Iterate backward. Start with the bottom row & move our way up.
for z in range(len(self.layers)-1, -1):
    # Get all entities on this layer.
    entities = self.entities[z]
    # Sort the entities by Y-position. FYI, sorted() returns a copy of the original list.
    entities = sorted(by_y, entities)

    for y in range(self.height):
        # Draw a row of tiles.
        for x in range(self.width):
            draw_tile(x, y, z)

        # Draw entities on this row.
        while len(entities) > 0 and entities[0].y >= y:
            entities[0].draw()
            entities.shift()

    # Pump remaining entities. They are all above the first tile row.
    while len(entities) > 0:
        entities[0].draw()
        entities.shift()
```

This way, tiles that are higher up are drawn over the player.

**Comment by seisatsu** [comment]
Try walking along the side of the building toward the roof in the testing world to see where the problem is a problem.

**Comment by pmer** [comment]
What's the relationship between this issue and #137?

My intuition says that fixing #137 would also fix the roof problem.

**Comment by pmer** [comment]
Do you think “pass through” tiles will provide a solution to the side/roof behavior?

**Comment by seisatsu** [comment]
Ok, on second thought, pass-through tiles will not fix the roof behavior. You are right that fixing #137 will fix this issue, rather this kind of situation just won't happen anymore with smart pixel-based collisions. Best to just fix #137.

**Comment by pmer** [comment]
Closing this issue as a duplicate of #137.
